File: Design-Patterns-Using-Go/Basics OOPS/Inheritance.go
package main

import "fmt"

// Person struct
type Person struct {
	name   string
	age    int
	gender string
}

// NewPerson constructor
func NewPerson(name string, age int, gender string) *Person {
	return &Person{name: name, age: age, gender: gender}
}

// String method for Person
func (p *Person) String() string {
	return fmt.Sprintf("Person: %s is a %s and %d years old.", p.name, p.gender, p.age)
}

// GetGender method for Person
func (p *Person) GetGender() string {
	return p.gender
}

// Citizen struct
type Citizen struct {
	Person
	id int
}

// NewCitizen constructor
func NewCitizen(name string, age int, id int, gender string) *Citizen {
	return &Citizen{Person: Person{name: name, age: age, gender: gender}, id: id}
}

// VoterID method for Citizen
func (c *Citizen) VoterID() int {
	return c.id
}

// String method for Citizen
func (c *Citizen) String() string {
	return fmt.Sprintf("Citizen: %s is a %s and %d years old with voter id %d.", c.name, c.gender, c.age, c.id)
}

// Client Code
func main() {
	p := NewPerson("John", 32, "Male")
	fmt.Println(p)

	c := NewCitizen("Smith", 31, 1234, "Male")
	fmt.Println(c)
}

/*
Person: John is a Male and 32 years old.
Citizen: Smith is a Male and 31 years old with voter id 1234.
*/

File: Design-Patterns-Using-Cpp/Basics OOPS/Inheritance.cpp
#include <iostream>
#include <string>

class Person {
protected:
    std::string name;
    int age;
    std::string gender;

public:
    Person(const std::string& name, int age, const std::string& gender)
        : name(name), age(age), gender(gender) {}

    virtual std::string toString() const {
        return "Person: " + name + " is a " + gender + " and " + std::to_string(age) + " years old.";
    }

    std::string getGender() const {
        return gender;
    }
};

class Citizen : public Person {
private:
    int id;

public:
    Citizen(const std::string& name, int age, int id, const std::string& gender)
        : Person(name, age, gender), id(id) {}

    int getVoterId() const {
        return id;
    }

    std::string toString() const override {
        return "Citizen: " + name + " is a " + gender + " and " + std::to_string(age) +
               " years old with voter id " + std::to_string(id) + ".";
    }
};

int main() {
    Person p("John", 32, "Male");
    std::cout << p.toString() << std::endl;

    Citizen c("Smith", 31, 1234, "Male");
    std::cout << c.toString() << std::endl;

    return 0;
}

/*
Person: John is a Male and 32 years old.
Citizen: Smith is a Male and 31 years old with voter id 1234.
*/